### ISSUE TITLE
Riak Returns Wrong Value for Key after PB request timeout happens [JIRA: CLIENTS-710]

### ISSUE THREAD

**Issue by ellulpatrick** [basho/riak-python-client#425]
This bug happens with the following flow.

A user gets a Connection object from the pool
1) sends Request X
2) waits for Response X
3) timeout happens, and dealt with at the application level

At this stage, the connection is released back in the pool, the next user comes along.
4) sends Request Y
6) receives Response X, because it was received in the meanwhile.

Some cases, the library throws: "'unexpected protocol buffer message code" if the requests were of different kinds.

Worst case, if the requests were both of the type "Get" but for a different key, you will ask for key X, but get value of Y.

We ended up with some corrupted data because of this.

To reproduce:

```
(2.7)➜  ~  virtualenv ~/riakbug
(2.7)➜  ~  source ~/riakbug/bin/activate
(riakbug)➜  ~  pip install riak
(riakbug)➜  ~  pip freeze
...
riak==2.3.0
...

(riakbug)➜  ~  python
Python 2.7.10 (default, Jul 14 2015, 19:46:27)

>>> from riak import RiakClient

>>> rc = RiakClient(protocol='pbc', host="172.17.8.101", pb_port=8087, transport_options={"timeout":0.01})
# you might have to play around with the timeout value to get the conditions right

>>> rc.ping()
True

>>> rc.get_buckets()
Traceback (most recent call last):
....
socket.timeout: timed out

>>> rc.bucket("a_bucket").get("a_key")
...
riak.riak_error.RiakError: 'unexpected protocol buffer message code: 16, <riak.pb.riak_kv_pb2.RpbListBucketsResp object at 0x10d442f50>'
# NOTE: that this is the response to the first request to get_buckets.

>>> rc.bucket("a_bucket").get("another_key")
<riak.riak_object.RiakObject object at 0x10d440b10>
# NOTE: This is the value of a_key, not of another_key

```

What could possibly be done about it is that when a timeout happens, the connection could be marked as a BadResource, and hence removed from the connection pool.
Another option for this specific problem is to flush any incoming data before making a new request.

This same issue could possibly also happen when other errors are thrown in step 3 above, not just for socket.timeout.

**Comment by lukebakken** [comment]
Thank you for the detailed report.

I'm assuming this is a rare event since only long-running Riak operations would time out in this way (such as list buckets in your example). Are you frequently listing buckets or running other expensive operations against your Riak cluster?

**Comment by ellulpatrick** [comment]
Hi Luke. We don't run any of the expensive operations in production. The
timeout must have occurred for a different reason which possibly we have no
control upon. Perhaps the underlying network, vm or hardware glitched
briefly. This is why we set a timeout of a couple of seconds. It has
happened a few times over the month or so we have been running the system
and we intend to patch it one way or another before continuing the rollout.

**Comment by ellulpatrick** [comment]
Just to clarify, the reproduction steps above are not our production code. We do not use get_buckets() in production. The steps above have a very low timeout and call a slower function on purpose to trigger the issue.
In production, in the dozen times that the issue happened, the timeout of 5s occurred on routine gets, puts, and searches.
We are working in a cloud/vm/container environment so it is probable that over time, we do get a random timeout for an uncontrollable reason. We handle this at the application layer, but we need the riak python connection pool to recover from it as well.
Thanks for your time.

**Comment by mcobzarenco** [comment]
+1

**Comment by lukebakken** [comment]
@mcobzarenco did you `+1` because you see this too?

**Comment by lukebakken** [comment]
Fixed by #451 and has been in releases since `2.5.2`

**Comment by Basho-JIRA** [comment]
Fixed, or closed via GitHub issues. 

***[posted via JIRA by Alexander Moore]***
